[SelectionDAG] Fix AArch64 machine verifier bug when expanding LOOP_DEPENDENCE_MASK

AZero13 opened the pull request:
Fixes: https://github.com/llvm/llvm-project/issues/168227

TargetConstant nodes don't match TableGen ImmLeaf patterns during instruction selection. When this zero constant flows into the AArch64 CCMP formation code, the machine verifier hits an assertion in expensive checks.

AZero13 committed:
Pre-commit test (NFC)

AZero13 commented:
@juliannagele

topperc commented:
The type legalizer changes LGTM. I don't know anything about the AArch64ISelLowering changes.

MacDue commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -26013,9 +26013,12 @@ static SDValue reassociateCSELOperandsForCSE(SDNode *N, SelectionDAG &DAG) {
+      std::swap(CmpOpToMatch, CmpOpOther);
+      if (SDValue R = Fold(NewCC, CmpOpToMatch, CmpOpToMatch))
+        return R;
+    }
```
Does this have any relation to the LOOP_DEPENDENCE_MASK change? Looks like you've just copied in https://github.com/llvm/llvm-project/pull/160753.

AZero13 replied:
Yeah that commit was fused in by mistake.

AZero13 committed:
[SelectionDAG] Fix AArch64 machine verifier bug when expanding LOOP_DEPENDENCE_MASK

Fixes: #168227

TargetConstant nodes don't match TableGen ImmLeaf patterns during instruction selection. When this zero constant flows into the AArch64 CCMP formation code, the machine verifier hits an assertion in expensive checks.
